src/net_processing.cpp

```diff
@@ -2579,8 +2579,23 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
             pfrom.fDisconnect = true;
             return;
         }
-        // ignore this for now - later we can downgrade the resources allocated
-        // to this peer.
+        if (pfrom.IsFullOutboundConn() || pfrom.IsAddrFetchConn()) {
+            // If we picked an outbound for tx- or addr-relay and it sends us a
+            // BLOCKRELAY, find another peer.
+            pfrom.fDisconnect = true;
+            return;
+        }
+        if (pfrom.m_tx_relay != nullptr && WITH_LOCK(pfrom.m_tx_relay->cs_filter, return pfrom.m_tx_relay->fRelayTxes)) {
+            // Can't send a BLOCKRELAY message if they didn't turn off tx-relay
+            // in the version message.
```

Comment:
This is where I don't understand BIP following excerpt :

> this is because the transaction relay field in the version message is not a permanent setting for the lifetime of the connection. Consequently, a node receiving an inbound connection with transaction relay disabled cannot distinguish between a peer that will never enable transaction relay and one that will.

Why do you mean exactly by "is not a permanent setting" ? Version message can only be received once (L2270). Further a peer which turns off tx-relay but then starts to relay tx would break its connection negotiation commitment and as such should be disconnected ? I don't think we do this for now. 

And even further when we pick up a peer as a block-relay-only one, we have no word to say in its choice of turning off tx-relay field or not ?

Reply:
Sending a FILTERCLEAR (at least, to a node with NODE_BLOOM set) will enable transaction relay; see BIP37.